### PR TITLE
request barrier  (testcafe part of the hammerhead #498 issue)

### PR DESCRIPTION
**Pull request by miherlosev**
@AlexanderMoskovkin @helen-dikareva
For more details see DevExpress/testcafe-hammerhead#498

### CODE REVIEW

File: src/client/core/request-barrier.js [hunk 1]
```diff
@@ -30,46 +30,61 @@ export default class XhrBarrier extends EventEmitter {
     }
 
     _init () {
-        var onXhrSend      = e => this._onXhrSend(e.xhr);
-        var onXhrCompleted = e => this._onXhrCompleted(e.xhr);
-        var onXhrError     = e => this._onXhrError(e.xhr, e.err);
+        var onXhrSend      = e => this._onRequestSend(e.xhr);
```
Comment (helen-dikareva) [inline src/client/core/request-barrier.js]: rename vars too

File: src/client/core/request-barrier.js [hunk 2]
```diff
@@ -30,46 +30,61 @@ export default class XhrBarrier extends EventEmitter {
...
                 this.emit(REQUESTS_FINISHED_EVENT);
         }
     }
 
+    _onFetchSend (request) {
+        if (this.collectingReqs) {
+            this.requests.push(request);
+
+            var onFetchRequestCompleted = () => this._onRequestCompleted(request);
+            var onFetchRequestError     = () => this._onRequestError(request);
```
Comment (helen-dikareva) [inline src/client/core/request-barrier.js]: I think we don't need this vars

File: src/client/core/request-barrier.js [hunk 3]
```diff
@@ -30,46 +30,61 @@ export default class XhrBarrier extends EventEmitter {
...
         this._unbindHandlers = () => {
             hammerhead.off(hammerhead.EVENTS.xhrSend, onXhrSend);
             hammerhead.off(hammerhead.EVENTS.xhrCompleted, onXhrCompleted);
             hammerhead.off(hammerhead.EVENTS.xhrError, onXhrError);
+            hammerhead.off(hammerhead.EVENTS.fetchSend, onFetchSend);
         };
     }
 
-    _onXhrSend (xhr) {
+    _onRequestSend (request) {
```
Comment (AlexanderMoskovkin) [inline src/client/core/request-barrier.js]: I suppose we should leave the old name for these functions. They handle just xhrs events, so it's `_onXhrSend/Completed/Error`.
In our context `request` it's xhr+fetch

File: src/client/core/request-barrier.js [hunk 4]
```diff
@@ -30,46 +30,61 @@ export default class XhrBarrier extends EventEmitter {
...
+            removeItem(this.requests, request);
 
             if (!this.collectingReqs && !this.requests.length)
                 this.emit(REQUESTS_FINISHED_EVENT);
         }
     }
 
+    _onFetchSend (request) {
+        if (this.collectingReqs) {
+            this.requests.push(request);
```
Comment (AlexanderMoskovkin) [inline src/client/core/request-barrier.js]: call `this._onRequestSend(request)` instead of this

File: src/client/core/request-barrier.js (hunk 2 above)
Comment (AlexanderMoskovkin) [inline src/client/core/request-barrier.js]: agree. This looks better:
```js
request.then(() => this._onRequestCompleted(request))
    .catch(() => this._onRequestError(request))
```

File: test/client/fixtures/core/request-barrier-test.js [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+
+QUnit.testDone(function () {
+    iframeSandbox.off(iframeSandbox.RUN_TASK_SCRIPT, window.initIFrameTestHandler);
+});
+
+
+$(document).ready(function () {
+    module('xhr', function () {
+        asyncTest('waitPageInitialRequests', function () {
+                var completeReqCount       = 0;
```
Comment (helen-dikareva) [inline test/client/fixtures/core/request-barrier-test.js]: delete extra indent for this block

File: test/client/fixtures/core/request-barrier-test.js [hunk 6]
```diff
@@ -0,0 +1,466 @@
...
+            window.setTimeout(function () {
+                ok(callbackRaised);
+                start();
+            }, 2000);
+        });
+    });
+
+    if (window.fetch) {
+        module('fetch', function () {
+                asyncTest('waitPageInitialRequests', function () {
```
Comment (helen-dikareva) [inline test/client/fixtures/core/request-barrier-test.js]: delete extra indent for this block too

File: test/client/fixtures/core/request-barrier-test.js [hunk 7]
```diff
@@ -0,0 +1,466 @@
...
+                                return response.text();
+                            })
+                            .then(function () {
+                                secondRequestCompleted = true;
+                            });
+                    }, 100);
+                });
+
+                asyncTest('barrier - creating new iframe without src (B236650)', function () {
+                    var $iframe           = null;
```
Comment (helen-dikareva) [inline test/client/fixtures/core/request-barrier-test.js]: I think this test is redundant (Isn't  it the same like test above https://github.com/DevExpress/testcafe/pull/548/files#diff-5b3b1295f2f653bab8bc18410ccaba81R201)

File: src/client/core/request-barrier.js (hunk 1 above)
Comment (miherlosev) [inline src/client/core/request-barrier.js]: `onXhrSend`, `onXhrCompleted`, `onXhrError` are used only for xhr events.
`_onRequestSend`, `_onRequestCompleted`, `_onRequestError` are common internal functions.  Used for xhr and fetch.
So I think we should not rename vars.

File: test/client/fixtures/core/request-barrier-test.js [hunk 8]
```diff
@@ -0,0 +1,466 @@
...
+                    });
+
+                    window.setTimeout(function () {
+                        ok(callbackRaised);
+                        start();
+                    }, 2000);
+                });
+            });
+    }
+});
```
Comment (AlexanderMoskovkin) [inline test/client/fixtures/core/request-barrier-test.js]: I suppose some of the `fetch` tests are redundant. In fact we haven't changed any logic in `xhr-barrier`, we just add `fetch` event handlers. So I propose leave old tests and add just two for `fetch`: the first with success fetch and the second with failed fetch.

File: src/client/core/request-barrier.js (hunk 1 above)
Comment (AlexanderMoskovkin) [inline src/client/core/request-barrier.js]: agree with @miherlosev here

File: src/client/core/request-barrier.js (hunk 2 above)
Comment (helen-dikareva) [inline src/client/core/request-barrier.js]: And move `.then` to new line)

File: src/client/core/request-barrier.js (hunk 1 above)
Comment (helen-dikareva) [inline src/client/core/request-barrier.js]: Ok. I got it.

**Comment by helen-dikareva** [comment]
\r-

File: src/client/core/request-barrier.js (hunk 2 above)
Comment (miherlosev) [inline src/client/core/request-barrier.js]: Ok

File: src/client/core/request-barrier.js (hunk 4 above)
Comment (miherlosev) [inline src/client/core/request-barrier.js]: Fetch promise processing is placed inside braces.
So I cannot use it.

Comment (AlexanderMoskovkin) [inline src/client/core/request-barrier.js]: yes, you are right

File: test/client/fixtures/core/request-barrier-test.js (hunk 8 above)
Comment (miherlosev) [inline test/client/fixtures/core/request-barrier-test.js]: Ok

**Comment by miherlosev** [comment]
Waiting for https://github.com/DevExpress/testcafe-hammerhead/issues/602

**Comment by inikulin** [comment]
@miherlosev Seems like there was a problem with rebase. I suggest you to rebase locally and open new pull request.

**Comment by miherlosev** [comment]
Yes. I will do it after merge https://github.com/DevExpress/testcafe-hammerhead/pull/607

**Comment by miherlosev** [comment]
I opened the another pull request - https://github.com/DevExpress/testcafe/pull/570

**Pull request closed without merging**